patches/api/0412-Player-Entity-Tracking-Events.patch

```diff
@@ -0,0 +1,102 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Yannick Lamprecht <yannicklamprecht@live.de>
+Date: Wed, 30 Mar 2022 18:16:37 +0200
+Subject: [PATCH] Player Entity Tracking Events
+
+
+diff --git a/src/main/java/io/papermc/paper/event/player/PlayerTrackEntityEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerTrackEntityEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..55a90b5f2be4edb01b042ed79b26fe7787d8c099
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/event/player/PlayerTrackEntityEvent.java
+@@ -0,0 +1,42 @@
++package io.papermc.paper.event.player;
++
++import org.bukkit.entity.Entity;
++import org.bukkit.entity.Player;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Is called before a {@link Player} tracks an {@link Entity}.
++ */
++public class PlayerTrackEntityEvent extends PlayerEvent {
++
++    private static final HandlerList handlers = new HandlerList();
++    private final Entity entity;
++
++    public PlayerTrackEntityEvent(@NotNull Player player, @NotNull Entity entity) {
++        super(player);
++        this.entity = entity;
++    }
++
++    @NotNull
++    public static HandlerList getHandlerList() {
++        return handlers;
++    }
++
++    @NotNull
++    @Override
++    public HandlerList getHandlers() {
++        return handlers;
++    }
++
++    /**
++     * Gets the entity that will be tracked
++     *
++     * @return the entity tracked
++     */
++    @NotNull
++    public Entity getEntity() {
++        return entity;
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/event/player/PlayerUntrackEntityEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerUntrackEntityEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..58e9ff364e5c5ea8983c8083b49b6dc3ba849ede
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/event/player/PlayerUntrackEntityEvent.java
+@@ -0,0 +1,42 @@
++package io.papermc.paper.event.player;
++
++import org.bukkit.entity.Entity;
++import org.bukkit.entity.Player;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.player.PlayerEvent;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++/**
++ * Is called before a {@link Player} untracks an {@link Entity}.
++ */
++public class PlayerUntrackEntityEvent extends PlayerEvent {
```

Comment:
Technically this is called after the entity is untracked, but BEFORE the packet to remove the entity is sent to the client.

Maybe for both of these just say "when X tracks/untracks entity"? I donno

Reply:
Also, why is Entity Nullable here? Doesn't make sense